src/impl/transact_log_handler.cpp

```diff
@@ -153,17 +162,34 @@ void KVOAdapter::before(SharedGroup& sg)
         auto& builder = list.builder;
         auto& changes = list.observer->changes[list.col];
 
+        builder.modifications.remove(builder.insertions);
+
         // KVO can't express moves (becaue NSArray doesn't have them), so
         // transform them into a series of sets on each affected index when possible
-        if (!builder.insertions.empty() && builder.insertions.count() == builder.moves.size()) {
+        if (!builder.moves.empty() && builder.insertions.count() == builder.moves.size() && builder.deletions.count() == builder.moves.size()) {
             changes.kind = BindingContext::ColumnInfo::Kind::Set;
             changes.indices = builder.modifications;
-
-            size_t start = std::min(builder.insertions.begin()->first, builder.deletions.begin()->first);
-            size_t end = std::max(std::prev(builder.insertions.end())->second,
-                                  std::prev(builder.deletions.end())->second);
-            for (size_t i = start; i < end; ++i)
-                changes.indices.add(i);
+            changes.indices.add(builder.deletions);
+
+            auto in_range = [](auto& it, auto end, size_t i) {
+                if (it != end && i >= it->second)
+                    ++it;
+                return it != end && i >= it->first && i < it->second;
+            };
+
+            auto del_it = builder.deletions.begin(), del_end = builder.deletions.end();
+            auto ins_it = builder.insertions.begin(), ins_end = builder.insertions.end();
+            size_t start = std::min(ins_it->first, del_it->first);
+            size_t end = std::max(std::prev(ins_end)->second, std::prev(del_end)->second);
+            ptrdiff_t shift = 0;
+            for (size_t i = start; i < end; ++i) {
+                if (in_range(del_it, del_end, i))
+                    --shift;
+                else if (in_range(ins_it, ins_end, i + shift))
+                    ++shift;
+                if (shift != 0)
+                    changes.indices.add(i);
+            }
```

Comment:
The KVO tests for move() and swap() all rely on this.

Previously move() just marked the full potentially-modified range as modified, and swap() marked the two relevant indices as modified (and didn't actually record that the rows were moved). However, this logic doesn't work for subtables, as Table::move_row() is emulated via insert/swap/erase and so for KVO it'd decay to just SetAll (because there's multiple kinds of changes). As a result I adjusted it so that swap() is actually handled as a pair of moves, which made the insert/swap/erase dance work, but also made it so that swap() would mark all the rows in between the swapped rows as modified, even though they didn't change. This code now enumerates over all of the rows which might have been shifted by moves and calculates if it actually has been shifted, or if the combination of moves has resulted in the row being in the same place as it started.